src/main/java/com/google/devtools/build/lib/buildeventservice/client/BuildEventServiceGrpcClient.java

```diff
@@ -48,11 +53,24 @@
   private final PublishBuildEventBlockingStub besBlocking;
 
   public BuildEventServiceGrpcClient(
-      ManagedChannel channel, @Nullable CallCredentials callCredentials) {
-    this(
-        withCallCredentials(PublishBuildEventGrpc.newStub(channel), callCredentials),
-        withCallCredentials(PublishBuildEventGrpc.newBlockingStub(channel), callCredentials),
-        channel);
+      ManagedChannel channel,
+      @Nullable CallCredentials callCredentials,
+      List<Map.Entry<String, String>> headers) {
+    ClientInterceptor interceptor = null;
```

Comment:
If the `BuildEventServiceGrpcClient` took the `ClientInterceptor` as a `@Nullable` constructor argument, then this class can remain focused on the business logic needed for a fully-constructed gRPC client.

The caller (`BazelBuildEventServiceModule`) can do the work of munging the configuration's key-value-pair headers into Metadata and a ClientInterceptor. In the future, if we need to change the Metadata or the ClientInterceptor (eg. adding tracing functionality), that change remains isolated to the Module.

This may also improve unit-testability of the `BuildEventServiceGrpcClient` by allowing tests to interpose the Interceptor on the client side.

Reply:
Ah, yes, I went back and forth myself on this issue when writing this CL. I agree that the munging would be more comfortable in `BazelBuildEventServiceModule`, but I will need some guidance to get there. Ultimately, I took uneasily took the current approach because

1. `BuildEventServiceGrpcClient` is already "polluted" by taking in `CallCredentials` and having to mess with that, so what more is some header configuration?
2. More pragmatically, I found it significantly easier to unittest `BuildEventServiceGrpcClient` than `BazelBuildEventServiceModule`. ``BazelBuildEventServiceModule` is basically a "plugin" for its superclass, and it looks tricky to unittest what I want without disturbing too much of that ecosystem. (There used to be `BazelBuildEventServiceModuleTest.java` until it slunk away in 4ca2b6d5c4d201009ff21a3abbeddd9d58d4c95b. I'm not sure what infrastructure exists for testing `BazelBuildEventServiceModule` within Google and don't want to reinvent it. Additionally, I see that `BazelBuildEventServiceModule#newGrpcChannel` has a comment explaining `newGrpcChannel is only defined so it can be overridden in tests to not use a real network link`, but no such overrides exist in the public test suite. This makes me nervous about doing surgery on this class without a full view of the codebase.)

Reply:
Thanks for your quick followup!

Re: #1: My thought here is that `CallCredentials` is a gRPC domain object, from the `io.grpc` package, and the `ClientInterceptor` would be also. So we'd be handing a few gRPC objects to construct a gRPC client, which gets to be focused entirely on gRPC logic.

Re: #2: It looks like an oversight that `BazelBuildEventServiceModuleTest.java` is not open-sourced. Thanks for pointing that out, I'll remedy that ASAP! If you want, you can pull down [the patchset](https://bazel-review.googlesource.com/c/bazel/+/175231) and work off of that. You could honestly just make the Flag-and-String-munging be a static method on `BazelBuildEventServiceModule` and drop some unit tests into `BazelBuildEventServiceModuleTest` that call the static method. We already have a `#testKeywords()` test in there for what amounts to testing Flag manipulation.

Reply:
Actually, it looks like open-sourcing `BazelBuildEventServiceModuleTest.java` will be a bit more work than I expected, so I'm going to approve this and merge it in.

Reply:
Alright,; thank you for your reviews and assistance on this.